torch/csrc/jit/script/compiler.cpp

```diff
@@ -36,115 +36,148 @@ using TypeTable = std::unordered_map<std::string, TypePtr>;
+    // so we take their intersection.
+    return RefinementSet(
+        unionList(true_refinements_, rhs.true_refinements_),
+        intersectList(false_refinements_, rhs.false_refinements_));
+  }
+  RefinementSet Or(const RefinementSet& rhs) const {
+    // if the result of an OR is true, either a & b could have been true,
+    // so we take the intersection of a.true_refinements & b.true_refinements.
+    // if the result is false, both a and b had to be false,
+    // so we take their union.
+    return RefinementSet(
+        intersectList(true_refinements_, rhs.true_refinements_),
+        unionList(false_refinements_, rhs.false_refinements_));
   }
 
-  c10::optional<TypeAndRange> getRefinement(const std::string& name) const {
-    const auto& maybe_mapping = mappings_.find(name);
-    if (maybe_mapping == mappings_.end()) {
-      return c10::nullopt;
-    }
-    return maybe_mapping->second;
-  }
-
-  // return the intersection of the values to type mappings between this
-  // types can be unified
-  void intersectRefinements(const Refinements& other) {
-    Refinements ret;
-    for (const auto& name_mapping : mappings_) {
-      const auto& name = name_mapping.first;
-      const auto& mapping = name_mapping.second;
-      if (auto other_mapping = other.getRefinement(name_mapping.first)) {
-        const auto maybe_unified_type =
-            unifyTypes(mapping.first, other_mapping->first);
-        if (maybe_unified_type) {
-          ret.setRefinement(
-              name, TypeAndRange(*maybe_unified_type, mapping.second));
-        }
-      }
-    }
-    mappings_ = std::move(ret.mappings_);
-  }
-
-  // return the union of the values to type mappings in a and b whose
-  // types can be unified
-  void unionRefinements(const Refinements& other) {
-    Refinements ret;
-    for (const auto& name_mapping : mappings_) {
-      const auto& name = name_mapping.first;
-      const auto& mapping = name_mapping.second;
-      TypePtr t_1 = mapping.first;
-      if (auto other_mapping = other.getRefinement(name_mapping.first)) {
-        TypePtr t_2 = other_mapping->first;
-        c10::optional<TypePtr> maybe_unified_type = c10::nullopt;
-        if (t_1->isSubtypeOf(t_2)) {
-          maybe_unified_type = t_1;
-        } else if (t_2->isSubtypeOf(t_1)) {
-          maybe_unified_type = t_2;
-        }
-        if (maybe_unified_type) {
-          ret.setRefinement(
-              name, TypeAndRange(*maybe_unified_type, mapping.second));
-        }
-      } else {
-        ret.setRefinement(name, mapping);
+  RefinementSet Not() const {
+    return RefinementSet(false_refinements_, true_refinements_);
+  }
+  const std::vector<Refinement> activeRefinements() const {
+    return true_refinements_;
+  }
+
+ private:
+  static bool sameVar(const Refinement& a, const Refinement& b) {
+    return a.identifier() == b.identifier();
+  }
+  static Refinements unionList(const Refinements& a, const Refinements& b) {
+    Refinements result = a;
```

Comment:
Could consider calling this `unionSet` now that we are enforcing uniqueness of each element